api/src/main/java/org/apache/cloudstack/api/command/user/volume/ChangeOfferingForVolume.java

```diff
@@ -0,0 +1,139 @@
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+package org.apache.cloudstack.api.command.user.volume;
+
+import com.cloud.event.EventTypes;
+import com.cloud.exception.ConcurrentOperationException;
+import com.cloud.exception.NetworkRuleConflictException;
+import com.cloud.exception.InsufficientCapacityException;
+import com.cloud.exception.ResourceAllocationException;
+import com.cloud.exception.ResourceUnavailableException;
+import com.cloud.offering.DiskOffering;
+import com.cloud.storage.Volume;
+import org.apache.cloudstack.api.APICommand;
+import org.apache.cloudstack.api.ApiConstants;
+import org.apache.cloudstack.api.ApiErrorCode;
+import org.apache.cloudstack.api.BaseAsyncCmd;
+import org.apache.cloudstack.api.Parameter;
+import org.apache.cloudstack.api.ResponseObject;
+import org.apache.cloudstack.api.ServerApiException;
+import org.apache.cloudstack.api.command.user.UserCmd;
+import org.apache.cloudstack.api.response.DiskOfferingResponse;
+import org.apache.cloudstack.api.response.VolumeResponse;
+import org.apache.cloudstack.context.CallContext;
+import org.apache.log4j.Logger;
+
+@APICommand(name = ChangeOfferingForVolume.APINAME,
+        description = "Change disk offering of the volume and also an option to auto migrate if required to apply the new disk offering",
+        responseObject = VolumeResponse.class,
+        requestHasSensitiveInfo = false,
+        responseHasSensitiveInfo = false,
+        since = "4.16")
+public class ChangeOfferingForVolume extends BaseAsyncCmd implements UserCmd {
+
+    public static final String APINAME = "changeOfferingForVolume";
+    public static final Logger s_logger = Logger.getLogger(ChangeOfferingForVolume.class.getName());
+    private static final String s_name = "changeofferingforvolumeresponse";
+
+    /////////////////////////////////////////////////////
+    //////////////// API parameters /////////////////////
+    /////////////////////////////////////////////////////
+
+    @Parameter(name = ApiConstants.ID, entityType = VolumeResponse.class, required = true, type = CommandType.UUID, description = "the ID of the volume")
+    private Long id;
+
+    @Parameter(name = ApiConstants.DISK_OFFERING_ID,
+            entityType = DiskOfferingResponse.class,
+            type = CommandType.UUID,
+            required = true,
+            description = "new disk offering id")
+    private Long newDiskOfferingId;
+
+    @Parameter(name = ApiConstants.SIZE, type = CommandType.LONG, required = false, description = "New volume size in GB")
+    private Long size;
```

Comment:
Does the Size and Iops parameters have to explicitly passed to the cmd with new offering?

Reply:
Yes @sureshanaparti, they have to pass explicitly only for custom disk offering. Updated the description to make it more clear. thanks